### PR TITLE
fix: config init -- Add retry when an invalid HW vendor is selected

### PR DESCRIPTION
**Pull request by LaVLaS**
During the HW vendor selection of `ilab config init`, add a retry when the user enters an invalid selection for the vendor and system profile.

CHANGE:
```
First, please select the hardware vendor your system falls into
[0] NO SYSTEM PROFILE
[1] AMD
[2] APPLE
[3] INTEL
[4] NVIDIA
Enter the number of your choice: 5                                                                                                                                                                                 
Error: 5 is not in the range 0<=x<=4.                                                                                                                                                                              
Enter the number of your choice: 4                                                                                                                                                                                 
You selected: NVIDIA

Next, please select the specific hardware configuration that most closely matches your system.                                                                                                                     
[0] NO SYSTEM PROFILE                                                                                                                                                                                              
[1] NVIDIA A100 X2                                                                                                                                                                                                 
[2] NVIDIA A100 X4                                                                                                                                                                                                 
[3] NVIDIA A100 X8                                                                                                                                                                                                 
[4] NVIDIA H100 X2                                                                                                                                                                                                 
[5] NVIDIA H100 X4                                                                                                                                                                                                 
[6] NVIDIA H100 X8                                                                                                                                                                                                 
[7] NVIDIA L4 X8                                                                                                                                                                                                   
[8] NVIDIA L40S X4                                                                                                                                                                                                 
[9] NVIDIA L40S X8                                                                                                                                                                                                 
Enter the number of your choice [hit enter for hardware defaults]: 10                                                                                                                                              
Error: 10 is not in the range 0<=x<=9.                                                                                                                                                                             
Enter the number of your choice [hit enter for hardware defaults]: -1                                                                                                                                              
Error: -1 is not in the range 0<=x<=9.                                                                                                                                                                             
Enter the number of your choice [hit enter for hardware defaults]: 11                                                                                                                                              
Error: 11 is not in the range 0<=x<=9.                                                                                                                                                                             
Enter the number of your choice [hit enter for hardware defaults]: 99                                                                                                                                              
Error: 99 is not in the range 0<=x<=9.                                                                                                                                                                             
Enter the number of your choice [hit enter for hardware defaults]: 9
```

**Issue resolved by this Pull Request:**
Resolves # [RHELAI-2399](https://issues.redhat.com/browse/RHELAI-2399)

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/config/init.py [hunk 1]
```diff
@@ -194,6 +194,12 @@ def prompt_user_to_choose_vendors(
         click.echo(
             "No profile selected - ilab will use generic code defaults - these may not be optimized for your system."
         )
+    else:
+        click.secho(
```
Comment (danmcp) [inline src/instructlab/config/init.py:198]: Would it be better to re-ask rather than exit?

Comment (LaVLaS) [reply src/instructlab/config/init.py:198]: That is a good suggestion.  Should that be part of another PR since that UX change to retry would need to be implemented in the vendor and hardware system selection.

Is there any requirement to fix this quickly for a `v0.21.z-stream` release and, if so, I can rebase this on top of `release-v0.21` branch then we can do another issues/PR for a retry option?  As it stands now the v0.21.z releases will reset the system profile config without warning if the user accidentally hits Enter or inputs an invalid number

Comment (danmcp) [reply src/instructlab/config/init.py:198]: No objection if we need to do a quick fix now and update later.  Re: 

> retry would need to be implemented in the vendor and hardware system selection

Not quite following this statement.  Wouldn't the loop be around the `click.prompt` just above?

Comment (LaVLaS) [reply src/instructlab/config/init.py]: `click.IntRange` did the trick

File: src/instructlab/config/init.py [hunk 2]
```diff
@@ -176,8 +176,7 @@ def prompt_user_to_choose_vendors(
 
     system_profile_selection = click.prompt(
         "Enter the number of your choice",
-        type=int,
-        default=0,
```
Comment (danmcp) [inline src/instructlab/config/init.py:180]: What happens if you set a default with IntRange also set?  Does is not still loop if you pick an invalid value?  It seems like was the original desired behavior.

File: src/instructlab/config/init.py [hunk 3]
```diff
@@ -190,10 +189,6 @@ def prompt_user_to_choose_vendors(
             )
         )
         click.echo("[0] NO SYSTEM PROFILE")
-    elif system_profile_selection == 0:
```
Comment (danmcp) [inline src/instructlab/config/init.py:193]: What was the reasoning for removing this message?  It seems unrelated to the change.

Comment (LaVLaS) [reply src/instructlab/config/init.py]: Apologies.  That is an error on my part.  I intended to remove the dead 'else:' code in `prompt_user_to_choose_profile()`

File: src/instructlab/config/init.py (hunk 2 above)
Comment (LaVLaS) [reply src/instructlab/config/init.py:180]: This may be an overreach on my part but removing the default requires the user to be explicit in the HW family choice.

Comment (danmcp) [reply src/instructlab/config/init.py:180]: I don't have a strong opinion.  Wanted to note it is a behavior change.

Comment (LaVLaS) [reply src/instructlab/config/init.py:180]: I restored the `default=0` after considering upstream/downstream documentation would need to be updated for that minor change.

File: src/instructlab/config/init.py [hunk 4]
```diff
@@ -205,7 +205,7 @@ def prompt_user_to_choose_profile(arch_family_processors) -> Config | None:
     cfg = None
     system_profile_selection = click.prompt(
         "Enter the number of your choice [hit enter for hardware defaults]",
-        type=int,
+        type=click.IntRange(0, len(arch_family_processors)),
```
Comment (danmcp) [inline src/instructlab/config/init.py:208]: nit: any reason why the above one uses `min=` and `max=` and this one doesn't?

**Review by danmcp — APPROVED** [review]

File: src/instructlab/config/init.py (hunk 4 above)
Comment (LaVLaS) [reply src/instructlab/config/init.py:208]: No real reason other than the 2nd was typed from memory.

Updated to remove min/max.

**Review by nathan-weinberg — APPROVED** [review]

**Comment by nathan-weinberg** [comment]
@Mergifyio rebase
